tests/Unit/Models/VoucherModelTest.php

```diff
@@ -204,4 +205,131 @@ public function testScopeConfirmedVouchers()
         $this->assertCount(3, Voucher::confirmed()->get());
         $this->assertEquals([2,3,4], Voucher::confirmed()->pluck('id')->toArray());
     }
+
+    /**
+     * Here because I can't work out how to test the Stateable Trait well
+     *
+     * @test
+     */
+    public function testItCanCreateAValidTransitionDefinition()
+    {
+        $validTransDef = Voucher::createTransitionDef("ordered", "print");
+
+        $this->assertEquals('print', $validTransDef->name);
+        $this->assertEquals('ordered', $validTransDef->from);
+        $this->assertNotNull('printed', $validTransDef->to);
+
+        // Has an invalid transition name
+        $this->assertNull(Voucher::createTransitionDef("ordered", "spacejam!"));
+
+        // Has an invalid transition "from" state
+        $this->assertNull(Voucher::createTransitionDef("kensington", "printed"));
+    }
+
+    /** @test */
+    public function testItCanCreateARangeDef()
+    {
+        // Make some vouchers
+        $sponsor = factory(Sponsor::class)->create([
+            'shortcode' => "TST",
+        ]);
+        $vouchers[] = factory(Voucher::class)->create([
+            'code' => 'TST0010',
+            'sponsor_id' => $sponsor->id,
+        ]);
+        $vouchers[] = factory(Voucher::class)->create([
+            'code' => 'TST1000',
+            'sponsor_id' => $sponsor->id,
+        ]);
+
+        // Make some input
+        $validRangeDef = Voucher::createRangeDefFromVoucherCodes(reset($vouchers)->code, end($vouchers)->code);
+
+        $this->assertEquals($sponsor->id, $validRangeDef->sponsor_id);
+        $this->assertInternalType('integer', $validRangeDef->sponsor_id);
+        $this->assertEquals($sponsor->shortcode, $validRangeDef->shortcode);
+        $this->assertEquals(10, $validRangeDef->start);
+        $this->assertInternalType('integer', $validRangeDef->start);
+        $this->assertEquals(1000, $validRangeDef->end);
+        $this->assertInternalType('integer', $validRangeDef->end);
+
+        // If you pass it a duff shortcode, it takes exception
+        $this->expectException(ModelNotFoundException::class);
+        Voucher::createRangeDefFromVoucherCodes('INV999998', 'INV999999');
+    }
+
+    /** @test */
+    public function testItCanFindASupersetRangeFromARangeSet()
+    {
+        $rangeCodes = [
+            'tst0101',
```

Comment:
Why using lowercase sponsor code?

Reply:
ooh. shouldnt be, can cause an issue. have fixed.